using_restricted_array.rb

```diff
@@ -7,41 +7,122 @@
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
 def length(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    array[i]
+    i += 1
+  end
+  return i
 end
 
 # Prints each integer values in the array
 def print_array(array)
-  puts "NOT IMPLEMENTED"
+  i = 0
+  until array[i] == nil
+    puts array[i]
+    i += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
 def search(array, length, value_to_find)
-  puts "NOT IMPLEMENTED"
+  if length == 1
+    return array[0] == value_to_find
+  else
+    i = 0
+    while i < length
+      if array[i] == value_to_find
+        return true
+      else
+        i += 1
+      end
+    end
+    return false
+  end
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
 def find_largest(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
+    return nil
+  elsif length == 1
+    return array[0]
+  else
+    i = 0
+    greatest = array[0]
+    while i < length
+      if array[i] > greatest
+        greatest = array[i]
+      else
+        i += 1
+      end
+    end
+    return greatest
+  end
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
 def find_smallest(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
+    return nil
+  elsif length == 1
+    return array[0]
+  else
+    i = 0
+    smallest = array[0]
+    while i < length
+      if array[i] < smallest
+        smallest = array[i]
+      else
+        i += 1
+      end
+    end
+    return smallest
+  end
 end
 
 # Reverses the values in the integer array in place
 def reverse(array, length)
-  puts "NOT IMPLEMENTED"
+  if length == 0
```

Comment:
Since this is reversing in place, you needn't return anything (like `nil` or `array[0]`). Just `return` would do.
As an optimization, your loop later will take care of both scenarios (if length is 0 or 1, the loop won't be entered and the method will just return).